### PR TITLE
Implement PUT /users/{user_id}/password endpoint for APIv4

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Implement PUT /users/{user_id}/password endpoint for APIv4.

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (mattermost/mattermost-api-reference#79)
- [x] All new/modified APIs include changes to the drivers
- [x] Touches critical sections of the codebase (password update)

### CODE REVIEW

File: model/client4.go [hunk 1]
```diff
@@ -212,6 +212,17 @@ func (c *Client4) UpdateUser(user *User) (*User, *Response) {
...
 }
 
+// UpdateUserPassword updates a user's password.
+func (c *Client4) UpdateUserPassword(userId, currentPassword, newPassword string) (*User, *Response) {
+	requestBody := map[string]string{"current_password": currentPassword, "new_password": newPassword}
+	if r, err := c.DoApiPut(c.GetUserRoute(userId)+"/password", MapToJson(requestBody)); err != nil {
+		return nil, &Response{StatusCode: r.StatusCode, Error: err}
+	} else {
+		defer closeBody(r)
+		return UserFromJson(r.Body), BuildResponse(r)
```
Comment (enahum) [inline model/client4.go]: im I reading this wrong? the API says that it returns `ReturnStatusOK(w)` but the driver returns a `*User`?

Comment (jwilander) [inline model/client4.go]: Whoops, that's a mistake

**Comment by jwilander** [comment]
@enahum fixed

File: api4/user.go [hunk 2]
```diff
@@ -109,6 +110,38 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	if len(currentPassword) <= 0 {
+		c.SetInvalidParam("current_password")
+		return
+	}
+
+	newPassword := props["new_password"]
+
+	c.LogAudit("attempted")
+
+	if c.Params.UserId != c.Session.UserId {
```
Comment (crspeller) [inline api4/user.go]: What about the system admin?

Comment (jwilander) [inline api4/user.go]: This endpoint requires having the current password, which the admin won't have. There'll be an admin specific endpoint for updating a user's password like we have for APIv3

Comment (crspeller) [inline api4/user.go]: Why does the endpoint require a userid then? Shouldn't we just get it from the session?

Comment (jwilander) [inline api4/user.go]: Yeah we can do that, I just had it match the other partial updates. I'll remove the user ID from the endpoint

**Comment by jwilander** [comment]
@crspeller Gorilla mux was causing havoc with the regex when I tried to do it the first way. So instead of dealing with that, I left the endpoint the same and added the ability for the system admin to update a user's password through this endpoint instead of using a different endpoint for that.

Ready for re-review.

File: api4/user_test.go [hunk 3]
```diff
@@ -210,3 +210,75 @@ func TestUpdateUserRoles(t *testing.T) {
...
+	CheckBadRequestStatus(t, resp)
+	_, resp = Client.UpdateUserPassword(th.BasicUser.Id, "badpwd", "newpwd")
+	CheckBadRequestStatus(t, resp)
+
+	// Should fail because account is locked out
+	_, resp = Client.UpdateUserPassword(th.BasicUser.Id, th.BasicUser.Password, "newpwd")
+	CheckErrorMessage(t, resp, "api.user.check_user_login_attempts.too_many.app_error")
+	CheckForbiddenStatus(t, resp)
+
+	// System admin can update another user's password
```
Comment (crspeller) [inline api4/user_test.go]: Can you add at test for a non-admin trying to update another real user's password?

Comment (jwilander) [inline api4/user_test.go]: Already have a case for that here https://github.com/mattermost/platform/pull/5243/files#diff-e35a6b6a1158b47b99f695aa19e908aeR250

Comment (crspeller) [inline api4/user_test.go]: Missed that. Was looking at the users being acted on :P

**Pull request merged**